zerver/lib/rate_limiter.py

```diff
@@ -160,6 +164,106 @@ def rate_limit_entity(cls, entity_key: str, rules: List[Tuple[int, int]],
         # Returns (ratelimited, secs_to_freedom)
         pass
 
+class TornadoInMemoryRateLimiterBackend(RateLimiterBackend):
+    # reset_times[key][rule] is the time at which the event
+    # request from the rate-limited key will be accepted.
+    reset_times = {}  # type: Dict[str, Dict[Tuple[int, int], float]]
+
+    # last_gc_time is the last time when the garbage was
+    # collected from reset_times.
+    last_gc_time = time.time()
+
+    # timestamps_blocked_until[key] contains the timestamp
+    # up to which the key has been blocked manually.
+    timestamps_blocked_until = {}  # type: Dict[str, float]
+
+    @classmethod
+    def need_to_limit(cls, entity_key: str, time_window: int,
+                      max_count: int) -> Tuple[bool, float]:
+        '''
+        Returns a tuple of `(rate_limited, time_till_free)`.
+        For simplicity, we have loosened the semantics here from
+        - each key may make atmost `count * (t / window)` request within any t
+          time interval.
+        to
+        - each key may make atmost `count * [(t / window) + 1]` request within
+          any t time interval.
+        Thus, we only need to store reset_times for each key which will be less
+        memory-intensive. This also has the advantage that you can only ever
+        lock yourself out completely for `window / count` seconds instead of
+        `window` seconds.
+        '''
+        now = time.time()
+
+        # Remove all timestamps from `reset_times` that are too old.
+        if cls.last_gc_time <= now - time_window / max_count:
+            cls.last_gc_time = now
+            cls.reset_times = {key: time_dict for (key, time_dict) in cls.reset_times.items()
+                               if time_dict[(time_window, max_count)] > now}
+
+        if entity_key in cls.timestamps_blocked_until:
+            # Check whether the key is manually blocked.
+            if now < cls.timestamps_blocked_until[entity_key]:
+                blocking_ttl = cls.timestamps_blocked_until[entity_key] - now
+                return True, blocking_ttl
+            else:
+                # Remove from timestamps_blocked_until if the blocking
+                # time has been expired.
+                del cls.timestamps_blocked_until[entity_key]
+
+        if entity_key in cls.reset_times:
+            # Check whether the new request would overfill its bucket.
+            new_reset = max(cls.reset_times[entity_key][(time_window, max_count)],
+                            now) + time_window / max_count
+        else:
+            # The bucket was empty, so one request cannot overfill it.
+            new_reset = now + time_window / max_count
+
+        if new_reset > now + time_window:
+            # Compute for how long the bucket will remain filled.
+            time_till_free = new_reset - time_window - now
+            return True, time_till_free
+
+        cls.reset_times[entity_key] = {(time_window, max_count): new_reset}
```

Comment:
I think we want `cls.reset_times[entity_key][(time_window, max_count)]` = new_reset`. Right now, since we're running this function for each rule, this will lose the reset times for the previous rules in the iteration?